primer/src/Primer/Action.hs

```diff
@@ -1290,7 +1309,7 @@ toProgActionInput def0 sel0 opt0 = \case
     opt <- optNoCxt
     sel <- typeSel
     index <- length . astTypeDefParameters <$> typeDef -- for now, we always add on to the end
-    pure [AddTypeParam sel.def index opt C.KType]
+    pure [AddTypeParam sel.def index opt $ C.KHole ()]
```

Comment:
Good question. I'm not sure.

There is one sense in which kinds are different to terms/types: since they are so simple, there is a canonical thing to put in the leaves of a non-holey tree: a `KType`. Since our kind actions will all push the current kind down (i.e. `k ~> k -> ?`), one will always end up filling in this hole with a `KType`. (This is not true in types/terms, since there are more options for leaves.).

Indeed, if we are going the direction of "leave as `KType` because of the above justification", I would perhaps argue that inserting an arrow should not create `k -> ?`, but `k -> KType`.

Reply:
Interesting. One could argue that we should go a step further and not allow constructing `KHole` at all. In practice I suppose this would mean, in addition to reverting the change above, replacing the "delete kind" action with a "raise" action, offered only on `KFun` nodes.

Now that I think about it, the purpose of `KHole` is somewhat unclear, given that we have such a simple kind system. Is it just used to give a kind to `TEmptyHole`? And would it be reasonable to just say that that has the kind `KType` instead?

Reply:
> Interesting. One could argue that we should go a step further and not allow constructing `KHole` at all. In practice I suppose this would mean, in addition to reverting the change above, replacing the "delete kind" action with a "raise" action, offered only on `KFun` nodes.

Potentially, but see below about why `KHole` exists (and is not easily going away). Thus if we forbade constructing `KHole` at all, it would be somewhat inconsistent.

Being able to give hole kinds does enable students to write more "programs" (i.e. things the system accepts, but are obvious nonsense). This may be good or bad depending on whether this is the sort of confusion/freedom that is good for learning(!), and on whether these states actually have utility as intermediate states in interactive use (I have not got any intuition on this as yet). In particular, one could write something like `data T (p : ?) = C (p p) p`, where `p` is being used at multiple kinds: (roughly speaking) `k -> *`, `k` and `*` (for some unknown `k`). It is not possible to fill in the `KHole` here to get a valid complete program. (As an aside, it is possible to make sense of this in a richer type system, by giving `p` a polymorphic kind; e.g. ghc will accept `data D (p :: forall k . k) = C p (p p)`. I don't know whether things of this ilk are practically useful.)

> Now that I think about it, the purpose of `KHole` is somewhat unclear, given that we have such a simple kind system. Is it just used to give a kind to `TEmptyHole`? And would it be reasonable to just say that that has the kind `KType` instead?

It is indeed used to kind `TEmptyHole` (iirc). In the past we did say `TEmptyHole` has kind `KType`, but there was an issue with that, and I changed it https://github.com/hackworthltd/vonnegut/commit/0f291a3db0904e1d05b1af60201358565089ac9a . In particular, we really want to be able to create type applications via the sequence (e.g.) `?  ;   ? ?   ;   List ?  ;  List Bool`, but the intermediate type `? ?` is checked by synthesising a kind for the first hole, ensuring it is (consistent with) an arrow kind and the second hole matches the domain -- this would fail if we gave the kind `KType`. (This is not the only way to solve this problem, but is the one we currently use, and would need to revisit this if we wanted to give the kind `KType` to type holes.)